### PR TITLE
Babel 7, add tree shaking, replace latest with env

### PR DESCRIPTION
**Pull request by EmilTholin**
This PR adds tree shaking and replaces the deprecated babel preset `latest` with `env`.

There is a slight misconfiguration of the `latest` preset:

```javascript
["latest", {
  "loose": true,
  // "modules": false // This shall be configured in "es2015" below
  // This will make it so no modules are transformed.
  "es2015": {
    "modules": false
  }
}]
```

Not transforming the modules to `commonjs` sets the code up nicely to be tree shaken, but the code still has to be transformed in order to run in Node. I therefore moved the `.babelrc` to a custom preset `.babelrc.js`, so `modules` could be conditionally turned on and off depending on if the code shall be used by Node or not.

Babel works by taking the closest `.babelrc` of the file it tries to import, so I turned of this feature with `babelrc: false` as an argument to `babel-register`, so that `static.config.js` is not loaded with the user's own babel config.

Webpack was now properly tagging parts that could be shaken off (e.g. `/* unused harmony export square */`), but they were still present in the resulting minified bundle. It turns out that `react-hot-loader/babel` makes tree shaking impossible, so removing that plugin for production made tree shaking finally work.

There are some exciting additional features in the `env` preset for babel 7, but I think it's best to wait with that until [react-hot-loader is up and running for babel 7](https://github.com/gaearon/react-hot-loader/pull/693).

### CODE REVIEW

**Comment by tannerlinsley** [comment]
This looks great. Are there any negative side effects from using preset instead of extend in project Babel rc?

**Comment by EmilTholin** [comment]
Thanks!
It was not possible to have `react-static` use `isNode` and still extend it in the examples. This works just as good. Do you know of any other nice way to achieve it?

**Comment by SShrike** [comment]
Shouldn't this be using `stage-2`, not `stage-0` as per the previous discussion?

**Comment by EmilTholin** [comment]
`stage-0` will only add the additional options to use `transform-do-expressions`,
`transform-function-bind` and `transform-export-extensions`, and will not clutter the bundle is they are not used, so I don't see an issue with it. I can change to `stage-2` if you like.

**Comment by SShrike** [comment]
That seems reasonable enough, I'll cede to you for the moment 😉.

File: .babelrc [hunk 1]
```diff
@@ -1,38 +1,5 @@
...
-        "stage-0",
-        "stage-3",
-        "react"
-      ]
-    }
-  }
-}
+    ["./.babelrc.js", { "isNode": true }]
+  ]
+}
```
Comment (SShrike) [inline .babelrc]: World's smallest nitpick: You're missing the final newline on this file.

**Comment by SShrike** [comment]
Also I never said this yet, but great work!

File: .babelrc (hunk 1 above)
Comment (EmilTholin) [reply .babelrc]: Ah, well spotted.

**Comment by EmilTholin** [comment]
Thanks!

**Comment by tannerlinsley** [comment]
@EmilTholin, it does work well, I was hoping not to perform any breaking changes for this upgrade though ;)

**Comment by EmilTholin** [comment]
Totally understand. I tried to have `.babelrc` as just `{ "presets": ["./.babelrc.js"] }`, and add `--presets=./.babelrc.js` to `build` and `watch` in order to not change the examples, but that didn't quite work sadly. I can give it some more thought if you like.

**Comment by tannerlinsley** [comment]
If it's a breaking change that's okay. Just want to try and avoid it if
possible. Let me know if it's not doable?

**Comment by EmilTholin** [comment]
I understand. I can research it a bit more today after work. I will keep you updated on how it goes.

**Comment by EmilTholin** [comment]
I have researched this some more, and I think I've stumbled upon something awesome.

Instead of conditionally transforming modules to `commonjs` with `isNode`, we can throw `commonjs` out of the window entirely by using [@std/esm](https://github.com/standard-things/esm). This way both environments can use ES6 and the same `.babelrc`, so no breaking change will be introduced.

This also means that `react-static` itself will be shaken, because it is no longer `commonjs` modules that has to be included in its entirety.

**Comment by tannerlinsley** [comment]
Outstanding!!!

**Comment by EmilTholin** [comment]
I still need to test this more thoroughly. Shall I ping you when it feels solid?

**Comment by tannerlinsley** [comment]
Yeah, as soon as it's ready, we can put it through the ringer with some experimentation and existing sites.

**Comment by tannerlinsley** [comment]
@EmilTholin Any updates on this? I would be lying if I said I wasn't super stoked for this PR.

**Comment by EmilTholin** [comment]
It's going great! The last missing piece of the puzzle with [`@std/esm` and `@babel/register`](https://github.com/standard-things/esm/issues/143) is soon complete, and then Babel 7 will come with this PR as well.

**Comment by tannerlinsley** [comment]
Awesome. Can't wait!

**Comment by SShrike** [comment]
Have to say I'm quite impressed so far, probably a better solution than I would've ended up having done :p

**Comment by SShrike** [comment]
Babel v7.0.0-beta.32 is out which includes jdalton's patch to babel-register, so everything should be all set at this point right?

**Comment by EmilTholin** [comment]
@SShrike Yeah, the patch to [Pirates](https://github.com/ariporad/pirates) needed to be released, and then [esm](https://github.com/standard-things/esm) as well. `esm` got a new release yesterday, so we should be set! I will add it after work today.

**Comment by EmilTholin** [comment]
This PR now also includes Babel 7.

[react-hot-loader](https://github.com/gaearon/react-hot-loader/blob/master/CHANGELOG.md) got support for Babel 7 in `3.1.2`, and it's working great! However, they removed `RedBox` as default error catcher in `3.0.0`, so the examples will no longer have that nice error in the browser, only in the console. The [`AppContainer` has to be changed to this to keep the functionality](https://github.com/gaearon/react-hot-loader#adding-a-custom-error-reporter).

[`@babel/preset-env` also got the super cool feature of `useBuiltIns: 'usage'` in Babel 7](https://github.com/babel/babel/tree/master/experimental/babel-preset-env#usebuiltins), but there is currently no way to customize the `moduleName`, so it looks for the polyfills relative to the user's folder and crashes, so this has to be added at a later stage (if at all).

Since both environments now use ES6, there is really no reason for the `watch` and `build` commands. `bin/react-static` could be changed to `require('../src/commands/index').default()` and `package.json.main` to `src/index.js`, but this would be a breaking change since e.g. `'react-static/lib/webpack/rules'` would be changed to `'react-static/src/webpack/rules'`, so that is probably best saved for the next major version (if it should be changed at all).

**Comment by tannerlinsley** [comment]
So can you confirm this is working for existing projects and fresh installs? (minus the old redbox behavior)

**Comment by EmilTholin** [comment]
Yep! I've tested thoroughly with `basic` and `dynamic-imports` and fresh installs. It would of course be great if you had the time to checkout the PR and play around with it as well.

**Comment by tannerlinsley** [comment]
Can do

**Comment by tannerlinsley** [comment]
I'm getting an error:
screen shot 2017-11-13 at 2 00 13 pm
Thoughts?

**Comment by tannerlinsley** [comment]
Ah. Looks like this needed to happen:
```javascript
// require.resolve('react-dev-utils/webpackHotDevClient'),
`${require.resolve('webpack-dev-server/client')}?/`,
```

**Comment by EmilTholin** [comment]
Ah, shoot. How did you get this error? I can't reproduce it. [This might give some clues](https://github.com/facebookincubator/create-react-app/issues/3332).

**Pull request merged**